content/blog/elm-at-noredink/index.md

```diff
@@ -0,0 +1,445 @@
+        , view = view
+        , notFound =
+            { bare = \flags -> notFoundPage flags
+            , initialized = \flags _ -> notFoundPage flags
+            }
+        , shouldSendViewPageEvent = (/=)
+        }
+```
+
+I think that the general lesson here is that just because Elm provides an
+`update` function with the shape `msg -> model -> (model, Cmd msg)` your
+own implementation doesn't need to. You might need more arguments, or
+different types, or even more return values.  As long as you convert the
+result of your function to match the API of `Browser.element` you can have
+any level of customization.
+
+## How to write Elm files
+
+We don't believe in the mantra:
+
+> Prefer small files
+
+Here is a little snapshot of some large files in our codebase:
+
+```bash
+2251 src/Page/Learn/GuidedDrafts/Main.elm
+1773 src/Page/Curriculum/BrowseAndAssignLayout.elm
+1725 src/Page/Teach/Classes/Page/Classes/Explore.elm
+1691 src/Page/Teach/Assignment/Form/GuidedDraft/Customize.elm
+1524 src/Page/Teach/Courses/PeerReviews/Progress.elm
+1280 src/Page/Admin/Alignments/Main.elm
+1188 src/Page/Learn/Home/Main.elm
+1184 src/Page/Teach/Assignment/Form/Model.elm
+1136 src/Page/Preview/PeerReview/Main.elm
+```
+
+As Evan explains in [this talk](https://www.youtube.com/watch/XpDsk374LDE),
+larger files in Elm are not a problem. We tend to delineate the different
+sections related to model, update and view with comments. Then we extract
+functions around their data structures, rather than their perceived role.
+
+Another point I want to emphasize is that in Elm you don't have to get
+everything right from the start. Choose the simplest architecture that
+works. If you need to change it later, the compiler will assist you in
+every step of the way. Specifically for this sort of mechanical changes,
+the chance of introducing bugs is very close to zero. If it compiles, it
+works.
+
+## Nesting the Elm Architecture
+
+When we need to combine multiple Elm applications, we nest them under one
+another. So for example if we have a modal in a page with its own state and
+its own messages, we will do this:
+
+```elm
+type alias Model =
+    { modal : Modal.Model
+    }
+
+type Msg = ModalMsg Modal.Msg
+
+update : Env -> Msg -> Model -> ( Model, Effect )
+update env msg model =
+    case msg of
+        ModalMsg modalMsg ->
+            let
+                ( newModal, modalEffect ) =
+                    Modal.update env modalMsg model.modal
+            in
+            ( { model | modal = newModal}
+            , ModalEffect modalEffect
+            )
+
+view : Env -> Model -> Html Msg
+view env model =
+    div []
+        [ Modal.view env model.modal |> Html.map ModalMsg
+        , span [] [ text "Hello there" ]
+        ]
+```
+
+This is how we build big apps in Elm. That's the secret sauce. By applying
+this concept repeatedly you can scale your Elm apps as much as you want.
+Here is the `Msg` type of our `Teach.Classes.Main` module:
```

Comment:
This seems to contradict the final takeaway of "use it as a last resort", and my reaction when I was first reading this was "weeeeell maybe we don't want to encourage folks to go straight to nested TEAs`.

I would try to put more emphasis on (or start with) "chosing the simplest solution that works", which IMO is the real secret sauce :-P

Maybe start with the simpler sidebar example and then say "but what if the component was complex state and we want to encapsulate that?".

Reply:
Since the article is full of great references, this might be a good place to link to Richard's [Scaling Elm Apps](https://www.youtube.com/watch?v=DoA4Txr4GUs&ab_channel=ElmEurope) talk.

Reply:
I think that my point I wanted to make is that you _will_ end up having nested TEAs to scale Elm apps, so being familiar with the idea is a must :D Then in some places you can avoid having it, but I felt it was untrue not to say that it's the tecnique we use the most.